From 852d4e72bfa920cedd7fa97b9fb9d5c8865c045e Mon Sep 17 00:00:00 2001
From: Zhou Guanghui <zhouguanghui1@huawei.com>
Date: Thu, 27 Oct 2022 02:46:26 +0000
Subject: [PATCH 062/595] [Huawei] mm: fix ignore cpuset enforcement

Offering: HULK
ascend inclusion
category: bugfix
bugzilla: 187895

-----------------------------------------------------

Since the current condition ignores the cpuset enforcement by adding
__GFP_THISNODEi to the gfp_mask, this will result in allocations that
specify __GFP_THISNODE and non-cdm nodes not subject to cpuset
restrictions.

For example, procA pid 1000:
node 0 cpus: 0 1 2 3
node 0 free: 57199MB
node 1 cpus: 4 5 6 7
node 1 free: 55930MB

cpuset/test/cpuset.mems  1
cpuset/test/tasks        1000
cpuset/test/cpuset.cpus  0-3

No cdm node exists. When procA malloc 100MB memory, the result is:
node 0 cpus: 0 1 2 3
node 0 free: 57099MB
node 1 cpus: 4 5 6 7
node 1 free: 55930MB
This is not what we expected, and in fact 100 MB of memory should be
allocated from node1. The reason for this problem is that THP specifies
__GFP_THISNODE to attempt to allocate from the local node.

Therefore, the cpuset enforcement should be ignored only when explicitly
allocating memory from the cdm node using __GFP_ THISNODE.

Fixes: b49907ccf1b4 ("mm: Ignore cpuset enforcement when allocation flag has __GFP_THISNODE")

Signed-off-by: Zhou Guanghui <zhouguanghui1@huawei.com>
---
 mm/page_alloc.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/mm/page_alloc.c b/mm/page_alloc.c
index c5bfc22e20d1..c63195b728a4 100644
--- a/mm/page_alloc.c
+++ b/mm/page_alloc.c
@@ -4997,7 +4997,8 @@ static inline bool prepare_alloc_pages(gfp_t gfp_mask, unsigned int order,
 	ac->migratetype = gfp_migratetype(gfp_mask);
 
 #ifdef CONFIG_COHERENT_DEVICE
-	if (cpusets_enabled() && !(*alloc_gfp & __GFP_THISNODE)) {
+	if (cpusets_enabled() &&
+	    (!(*alloc_gfp & __GFP_THISNODE) || !is_cdm_node(preferred_nid))) {
 #else
 	if (cpusets_enabled()) {
 #endif
-- 
2.17.1

